atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,223 @@
+
+
+class FetchSourcesPlugin(PreBuildPlugin):
+    """Download sources that may be used in further steps to compose Source Containers"""
+    key = PLUGIN_FETCH_SOURCES_KEY
+    is_allowed_to_fail = False
+    DOWNLOAD_DIR = 'image_sources'
+
+    def __init__(
+        self, tasker, workflow, koji_build_id=None, koji_build_nvr=None, signing_intent=None,
+        kojihub=None, kojiroot=None, koji_proxy_user=None, koji_ssl_certs=None, koji_principal='',
+        koji_keytab=''
+    ):
+        """
+        :param tasker: ContainerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param koji_build_id: int, container image koji build id
+        :param koji_build_nvr: str, container image koji build NVR
+        :param signing_intent: str, ODCS signing intent name
+        :param kojihub: str, koji hub (xmlrpc)
+        :param kojiroot: str, Koji root storage URL
+        :param koji_proxy_user: str, user to log in as (requires hub config)
+        :param koji_ssl_certs_dir: str, path to 'cert', 'ca', 'serverca'
+        :param koji_principal: str, Kerberos principal (must specify keytab)
+        :param koji_keytab: str, keytab name (must specify principal)
+        """
+        if not koji_build_id and not koji_build_nvr:
+            err_msg = ('{} expects either koji_build_id or koji_build_nvr to be defined'
+                       .format(self.__class__.__name__))
+            raise TypeError(err_msg)
+        type_errors = []
+        if koji_build_id is not None and not isinstance(koji_build_id, int):
+            type_errors.append('koji_build_id must be an int. Got {}'.format(type(koji_build_id)))
+        if koji_build_nvr is not None and not isinstance(koji_build_nvr, str):
+            type_errors.append('koji_build_nvr must be a str. Got {}'.format(type(koji_build_nvr)))
+        if type_errors:
+            raise TypeError(type_errors)
+
+        super(FetchSourcesPlugin, self).__init__(tasker, workflow)
+        self.koji_build = None
+        self.koji_build_id = koji_build_id
+        self.koji_build_nvr = koji_build_nvr
+        self.signing_intent = signing_intent
+        self.koji_fallback = {
+            'hub_url': kojihub,
+            'root_url': kojiroot,
+            'auth': {
+                'proxyuser': koji_proxy_user,
+                'ssl_certs_dir': koji_ssl_certs,
+                'krb_principal': koji_principal,
+                'krb_keytab_path': koji_keytab
+            }
+        }
+        self.session = get_koji_session(self.workflow, self.koji_fallback)
+
+    def run(self):
+        build_id = self.get_image_build_id()
+        signing_intent = self.get_signing_intent(build_id)
+        koji_config = get_koji(self.workflow, {})
+        insecure = koji_config.get('insecure_download', False)
+        urls = self.get_srpm_urls(build_id, signing_intent['keys'], insecure=insecure)
+        sources_list = self.download_sources(urls, insecure=insecure)
+        return sources_list
```

Comment:
Given how the build-step plugin wants to run, would it be nicer for this to return the output directory instead of all the files inside it? @MartinBasti wdyt?

Reply:
Oh yes, build_step plugin only needs path to directory with SRPMs

Reply:
Not right now. Without a scheme defined for what that should look like let's just stick to having everything at the top-level in this first iteration.